### DISCUSSION TITLE
rpc error: code = ResourceExhausted desc = trying to send message larger than max (140000321 vs. 104857600)

### DISCUSSION THREAD

**Discussion by xiaojuan8** [milvus-io/milvus#16092] [Q&A and General discussion]
Hi，I am testing search performance of my milvus cluster. I want to retrieve the performance of milvus benchmark : (https://milvus.io/cn/docs/v2.0.x/benchmark.md)， I set nq=10000 as config in benchmark. I use this nq for searching all the way. It works fine. But today when I just search with nq = 10000, error occurred as shown below : 

Failed to search: no Available QueryNode result, filter reason err: rpc error: code = ResourceExhausted desc = trying to send message larger than max (140000321 vs. 104857600)
, /go/src/github.com/milvus-io/milvus/internal/util/trace/stack_trace.go:51 github.com/milvus-io/milvus/internal/util/trace.StackTrace
/go/src/github.com/milvus-io/milvus/internal/util/grpcclient/client.go:253 github.com/milvus-io/milvus/internal/util/grpcclient.(*ClientBase).ReCall
/go/src/github.com/milvus-io/milvus/internal/distributed/proxy/client/client.go:146 github.com/milvus-io/milvus/internal/distributed/proxy/client.(*Client).SendSearchResult
/go/src/github.com/milvus-io/milvus/internal/querynode/proxy_session_manager.go:102 github.com/milvus-io/milvus/internal/querynode.(*SessionManager).SendSearchResult
/go/src/github.com/milvus-io/milvus/internal/querynode/query_collection.go:1396 github.com/milvus-io/milvus/internal/querynode.(*queryCollection).publishSearchResultWithCtx
/go/src/github.com/milvus-io/milvus/internal/querynode/query_collection.go:1400 github.com/milvus-io/milvus/internal/querynode.(*queryCollection).publishSearchResult
/go/src/github.com/milvus-io/milvus/internal/querynode/query_collection.go:1220 github.com/milvus-io/milvus/internal/querynode.(*queryCollection).search
/go/src/github.com/milvus-io/milvus/internal/querynode/query_collection.go:666 github.com/milvus-io/milvus/internal/querynode.(*queryCollection).doUnsolvedQueryMsg
/usr/local/go/src/runtime/asm_amd64.s:1371 runtime.goexit

: id 431882920772392797

**Comment by yhmo** [comment]
In the milvus.yaml, there are some configurations to define the rpc transfer limitation:

```
queryNode:
  grpc:
    serverMaxRecvSize: 2147483647 # math.MaxInt32
    serverMaxSendSize: 2147483647 # math.MaxInt32
    clientMaxRecvSize: 104857600 # 100 MB, 100 * 1024 * 1024
    clientMaxSendSize: 104857600 # 100 MB, 100 * 1024 * 1024
```

Change the clientMaxRecvSize/clientMaxSendSize to a lager values.

**Reply by JackyKwok01** [reply]
it seems we cant configure this setting in values.yaml. Have to override the config.tpl.

**Reply by yhmo** [reply]
> it seems we cant configure this setting in values.yaml. Have to override the config.tpl.

The DevOps will make an enhancement to [milvus-helm](https://github.com/milvus-io/milvus-helm) project to allow users to change the configurations more easily.

**Reply by mulakalagowthami** [reply]
![WhatsApp Image 2023-04-06 at 19 04 57](https://user-images.githubusercontent.com/125670012/230394625-ac5c3b6f-e4a6-4f5e-a08d-f00a819a669d.jpg)
I'm struggling with same problem can you please help me to resolve this

**Reply by yhmo** [reply]
The simple way to avoid this error is: **reduce the row count for each insert call.**
104857600 = 1GB, the upper limit of each insert call is 1GB.
Assume your vector dimensions is 512, each vector size is 512*4=2KB. 1GB/2KB=51200, which means you cannot insert to many vectors more than 51200 for each time.
Reducing the vector number for each insert call can avoid this error.

If you still encounter this error, show me the client code.

**Reply by mulakalagowthami** [reply]
I don't have any chaincode just I'm running test-network I'm new to this hyperledger fabric

**Reply by yhmo** [reply]
@mulakalagowthami 
I wrote a script to show why you got the error "ResourceExhausted desc = grpc: received message larger than max (xxxxxxx vs. 104857600)".

You can read the comment in this script, and run this script steps by steps, then you will understand why this error happens.

You can launch a [standalone milvus](https://milvus.io/docs/install_standalone-docker.md) before running this script.

```
import random
from pymilvus import (
    connections,
    FieldSchema, CollectionSchema, DataType,
    Collection,
    utility,
)

_HOST = '127.0.0.1'
_PORT = '19530'

if __name__ == '__main__':
    connections.connect(host=_HOST, port=_PORT)

    collection_name = "demo"
    if utility.has_collection(collection_name):
        utility.drop_collection(collection_name)

    # create a collection with these fields: id and vector
    dim = 512
    field1 = FieldSchema(name="id_field", dtype=DataType.INT64, is_primary=True, auto_id=True)
    field2 = FieldSchema(name="vector_field", dtype=DataType.FLOAT_VECTOR, dim=dim)
    schema = CollectionSchema(fields=[field1, field2])
    collection = Collection(name="demo", schema=schema)
    print("collection created")

    index_param = {
        "index_type": "FLAT",
        "params": {},
        "metric_type": "L2"}
    collection.create_index("vector_field", index_param)
    collection.load()

    # insert rows
    one_vector = [random.random() for _ in range(dim)]

    def do_insert(count):
        print("============================================================================")
        try:
            vectors = []
            for i in range(count):
                vectors.append(one_vector)
            data = [
                vectors,
            ]
            print(f"data {count} rows is ready")
            collection.insert(data)
            print(f"insert {count} rows successfully")
        except Exception as e:
            print(f"insert {count} rows failed, error:", e)
        print("============================================================================")

    # by default, the upper limit of data size sent to server side is 64MB
    # by default, the upper limit of result size returned to client side is 100MB
    # this threshold is defined in milvus.yml, both in the 'proxy' section and global(for the reason compatible with old version)
    # if you want to change the upper limit, change both of the configurations.
    # proxy:
    #     grpc:
    #         serverMaxRecvSize: 67108864  # 64M
    #         serverMaxSendSize: 67108864  # 64M
    #         clientMaxRecvSize: 104857600  # 100 MB, 100 * 1024 * 1024
    #         clientMaxSendSize: 104857600  # 100 MB, 100 * 1024 * 1024
    # grpc:
    #     serverMaxRecvSize: 67108864  # 64M
    #     serverMaxSendSize: 67108864  # 64M
    #     clientMaxRecvSize: 104857600  # 100 MB, 100 * 1024 * 1024
    #     clientMaxSendSize: 104857600  # 100 MB, 100 * 1024 * 1024

    # 1. now we test the upper limit(64MB) of sending data to server side by insert()
    # insert data size = dimension*sizeof(float32)*rows, in this script, the dimension is 512
    # insert 30000 rows, data size is 512*4*30000 = 60MB < 64MB, send data to server side successfully
    do_insert(count=30000)

    # insert 40000 rows, data size is 512*4*40000 = 80MB > 64MB, send data to server side failed
    # you will get error "grpc: received message larger than max (81920059 vs. 67108864)"
    do_insert(count=40000)


    def do_search(nq, topk):
        print("============================================================================")
        search_vectors = [[random.random() for _ in range(dim)] for _ in range(nq)]
        search_param = {
            "data": search_vectors,
            "anns_field": "vector_field",
            "param": {"metric_type": "L2", "params": {}},
            "limit": topk,
            "expr": "id_field > 0"}
        try:
            results = collection.search(**search_param)
            print(f"search nq={nq} topk={topk} successfully")
        except Exception as e:
            print(f"search nq={nq} topk={topk} failed, error:", e)
        print("============================================================================")


    # 2. now we test the upper limit(100MB) returned by search()
    # since search result is a list of pair id to distance, each pair size is sizeof(int64)+sizeof(float32)=12bytes
    # pair count = nq*topk, nq is the count of vectors you input to search().
    # then we know the total result size is nq*topk*12

    # search with nq=500, topk=10000, result size is 500*10000*12=60MB < 100MB, receive result successfully
    do_search(nq=500, topk=10000)

    # search with nq=1000, topk=10000, result size is 1000*10000*12=120MB > 100MB, receive result failed
    # you will get error "grpc: received message larger than max (130
```020047 vs. 104857600)"
    do_search(nq=1000, topk=10000)

**Comment by mulakalagowthami** [comment]
Thank you  but it is solved because I reinstalled all the fabric samples
again .
I definitely will run it.
I need to ask you one more thing the chaincodeinstallation installation on
peer is getting failed I don't why even though everything is correct.I
attached screenshot of it can please help me regarding that.

**Reply by yhmo** [reply]
What is the error?

**Comment by lukazso** [comment]
I am having the same issue that I cannot perform searches/queries without setting the limit very low; otherwise I receive the error mentioned above. However, I cannot find a `milvus.yaml` anywhere. In the current version, can someone point me to the right place where I can adjust the `clientMaxRecvSize` parameter accordingly? 

Thanks
Lukas

**Reply by xiaofan-luan** [reply]
I actually don't recommend you to change this. Usually means you are doing something wrong. By default the limit is 64M and it should be ok to hold more than 10K vectors.

if you really want to change the limit it's actually in the milvus.yaml.

**Reply by yhmo** [reply]
The milvus.yaml is inside the milvus container, under the path /milvus/configs/milvus.yaml
To change the configurations: https://milvus.io/docs/v2.2.x/configure-docker.md
